src/core/EffectChain.cpp

```diff
@@ -162,10 +162,10 @@ void EffectChain::removeEffect( Effect * _effect )
 
 void EffectChain::moveDown( Effect * _effect )
 {
-	if( _effect != m_effects.last() )
+	if (_effect != m_effects.back())
 	{
-		int i = m_effects.indexOf(_effect);
-		std::swap(m_effects[i + 1], m_effects[i]);
+		auto it = std::find(m_effects.begin(), m_effects.end(), _effect);
+		std::swap(*(it++), *it);
```

Comment:
Ah, I don't think so. It should be ``*(++it)`` I believe. However, it may be cleaner to just use ``std::distance`` and ``std::find`` together to get the index and go from there. Looking back, it's not so easy to tell if this works as intended.

Reply:
I mean is it going to affect the next `it` as well? We don't want to increment the pointer just point to the next one.

Reply:
This is most likely undefined behavior, since there is no way of knowing what arguments get evaluated first. I'll keep the original ``std::swap`` line and use ``std::distance`` and ``std::find`` to get the index instead.

Or, better yet, keep using ``it`` but with ``std::advance``.

Reply:
I'm rusty on pointers, why won't `*(it+1)` work?

Reply:
``std::find`` can be defined as (From the cppreference: https://en.cppreference.com/w/cpp/algorithm/find):
```
template< class InputIt, class T >
InputIt find( InputIt first, InputIt last, const T& value );
```

Evidently, the return type for ``std::find`` is of ``InputIt``, which must meet the named requirements of ``LegacyInputIterator``, so there is no guarantee that the iterator returned from ``std::find`` will support integral advances in this manner because a ``LegacyInputIterator`` can't be advanced this way (as seen here: https://en.cppreference.com/w/cpp/named_req/InputIterator). ``std::advance`` however, can advance any iterator fitting the requirements of ``LegacyInputIterator`` by an arbitary amount.

Reply:
In other words, ``std::find`` may return an iterator that meets the bare requirements of a ``LegacyInputIterator``, which technically cannot be advanced that way.

Reply:
> In other words, `std::find` may return an iterator that meets the bare requirements of a `LegacyInputIterator`, which technically cannot be advanced that way.

I think the template argument deduction will give `InputIt = std::vector<T>::iterator` which meets `LegacyRandomAccessIterator`, so itw won't be an issue.

If you're still worried, how about using [`std::next`](https://en.cppreference.com/w/cpp/iterator/next) and [`std::prev`](https://en.cppreference.com/w/cpp/iterator/prev) instead?

Reply:
> > In other words, `std::find` may return an iterator that meets the bare requirements of a `LegacyInputIterator`, which technically cannot be advanced that way.
> 
> I think the template argument deduction will give `InputIt = std::vector<T>::iterator` which meets `LegacyRandomAccessIterator`, so itw won't be an issue.
> 
> If you're still worried, how about using [`std::next`](https://en.cppreference.com/w/cpp/iterator/next) and [`std::prev`](https://en.cppreference.com/w/cpp/iterator/prev) instead?

``std::next`` and ``std::prev`` do seem like better choices in this case though rather than ``std::advance``, since they return the iterator that was incremented/decremented. Thanks for bringing them to my attention.

And it probably would be fine, its just that that may not always be the case, as to which it serves good practice and a reminder (atleast for me) to write code that works given various scenarios that won't always give back the same result.